drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -0,0 +1,426 @@
+                break;
+
+            case C7:
+                cfg->channels[15].pfpamsk = 1;
+            case A15:
+                cfg->channels[15].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 15;
+                break;
+
+            case C8:
+                cfg->channels[16].pfpamsk = 1;
+            case B0:
+                cfg->channels[16].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 16;
+                break;
+
+            case C9:
+                cfg->channels[17].pfpamsk = 1;
+            case B1:
+                cfg->channels[17].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 17;
+                break;
+
+            case C10:
+                cfg->channels[18].pfpamsk = 1;
+            case B2:
+                cfg->channels[18].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 18;
+                break;
+
+            case C11:
+                cfg->channels[19].pfpamsk = 1;
+            case B3:
+                cfg->channels[19].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 19;
+                break;
+
+            case C12:
+                cfg->channels[20].pfpamsk = 1;
+            case B4:
+                cfg->channels[20].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 20;
+                break;
+
+            case C13:
+                cfg->channels[21].pfpamsk = 1;
+            case B5:
+                cfg->channels[21].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 21;
+                break;
+
+            case C14:
+                cfg->channels[22].pfpamsk = 1;
+            case B6:
+                cfg->channels[22].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 22;
+                break;
+
+            case C15:
+                cfg->channels[23].pfpamsk = 1;
+            case B7:
+                cfg->channels[23].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 23;
+                break;
+        }
+    }
+}
+void rgb_callback(PWMDriver *pwmp);
+
+void shared_matrix_rgb_enable(void) {
+    pwmcfg.callback = rgb_callback;
+    pwmEnablePeriodicNotification(&PWMD1);
+}
+
+void shared_matrix_rgb_disable_pwm(void) {
+    // Disable PWM outputs on column pins
+    for(uint8_t y = 0; y < LED_MATRIX_COLS; y++) {
+        pwmDisableChannel(&PWMD1,chan_col_order[y]);
+    }
+}
+
+void shared_matrix_rgb_disable_leds(void) {
+    // Disable LED outputs on RGB channel pins
+    for (uint8_t x = 0; x < LED_MATRIX_ROWS_HW; x++) {
+        writePinLow(led_row_pins[x]);
+    }
+}
+
+void update_pwm_channels(PWMDriver *pwmp) {
+    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
+    for(uint8_t col_idx = 0; col_idx < LED_MATRIX_COLS; col_idx++, row_shifter <<= 1) {
+        #if(DIODE_DIRECTION == ROW2COL)
+            // Scan the key matrix column
+          matrix_read_rows_on_col(current_matrix,col_idx,row_shifter);
```

Comment:
Looks like this code performs a complete key matrix scan for every RGB row PWM cycle (unlike what happens for a `COL2ROW` matrix); is this really fast enough for good PWM performance? What is the resulting PWM frequency (the number of `rgb_callback()` calls per second), which would be equal to the actual scan rate in this configuration?

Reply:
This is a small issue I can't seem to get around. `ROW2COL` is slower in RGB performance. Not very much, but some flickering exists. Any pointers on updating the logic there appreciated